codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java

```diff
@@ -0,0 +1,108 @@
+
+final class Http2HeadersValidator {
+
+    private static final List<AsciiString> mandatoryHeaders = Collections.unmodifiableList(Arrays.asList(
+            PATH.value(), SCHEME.value(), METHOD.value()
+    ));
+
+    private static final List<AsciiString> connectionSpecificHeaders = Collections.unmodifiableList(
+            Arrays.asList(CONNECTION, TRANSFER_ENCODING, KEEP_ALIVE, UPGRADE));
+
+    private Http2HeadersValidator() {
+    }
+
+    /**
+     * Validates connection-specific headers according to
+     * <a href="https://tools.ietf.org/html/rfc7540#section-8.1.2.2">RFC7540, section-8.1.2.2</a>
+     */
+    static void validateConnectionSpecificHeaders(Http2Headers headers, int streamId) throws Http2Exception {
+        for (int i = 0; i < connectionSpecificHeaders.size(); i++) {
+            final AsciiString header = connectionSpecificHeaders.get(i);
+            if (headers.contains(header)) {
+                throw streamError(streamId, PROTOCOL_ERROR,
+                                  "Connection-speficic headers like [%s] must not be used with HTTP/2.", header);
+            }
+        }
+
+        final CharSequence teHeader = headers.get(TE);
+        if (teHeader != null && !AsciiString.contentEqualsIgnoreCase(teHeader, TRAILERS)) {
+            throw streamError(streamId, PROTOCOL_ERROR,
+                              "TE header must not contain any value other than \"%s\"", TRAILERS);
+        }
+    }
+
+    /**
+     * Validates Pseudo-Header fields according to
+     * <a href="https://tools.ietf.org/html/rfc7540#section-8.1.2.3">RFC7540, section-8.1.2.3</a>
+     */
+    static void validatePseudoHeaders(Http2Headers headers, int streamId) throws Http2Exception {
+        final CharSequence method = headers.get(METHOD.value());
+        if (method == null) {
+            throw streamError(streamId, PROTOCOL_ERROR,
+                              "Mandatory header [:method] is missing.");
+        }
+
+        if (HttpMethod.CONNECT.asciiName().contentEqualsIgnoreCase(method)) {
+            if (headers.contains(SCHEME.value())) {
+                throw streamError(streamId, PROTOCOL_ERROR,
+                                  "Header [:scheme] must be omitted when using CONNECT method.");
+            }
+
+            if (headers.contains(PATH.value())) {
+                throw streamError(streamId, PROTOCOL_ERROR,
+                                  "Header [:path] must be omitted when using CONNECT method.");
+            }
+        } else {
+            final CharSequence path = headers.get(PATH.value());
+            if (path != null && path.length() == 0) {
+                throw streamError(streamId, PROTOCOL_ERROR, "[:path] header cannot be empty.");
+            }
+
+            for (int i = 0; i < mandatoryHeaders.size(); i++) {
+                AsciiString header = mandatoryHeaders.get(i);
+                final int valueSize = headers.getAll(header).size();
```

Comment:
This is going to be relatively poor performing considering we're going to make a list for each mandatory header. Since the spec clearly defines which ones are mandatory could we just iterate through the header list in order and count them up, eg have `boolean methodObserved, pathObserved, ...` and make sure we don't do more than one and that we got all three?

Reply:
Iterating through the whole list would scale linearly with the number of headers whereas this version has a constant run time. If we assume the number of headers is most of the time small iterating would be more performant.

Ideally we would have a `count(name)` method on headers that would just count the headers with a given name, not convinced this usecase justify adding that to the API though.

Reply:
If we use my strategy of having an `observed` variable for each critical header we should count no more than 4 as at 4 we'd hit an exceptional condition. This actually gets to another conformance requirement of the RFC ([section 8.1.2.1](https://http2.github.io/http2-spec/#rfc.section.8.1.2.1)) which is that we should make sure we don't have bonus pseudo headers: `Pseudo-header fields defined for requests MUST NOT appear in responses; pseudo-header fields defined for responses MUST NOT appear in requests.`

Reply:
Nice, forgot about this requirement, will change it. 
(BTW 8.1.2.1 requirement is already checked in the HPackDecoder)

Reply:
They validate at different levels: in `HPackdecoder` we check if the headerBlock is valid, here we check if all the headers together make sense, that being said it would be nice if it was done at a single place.

The problem is that `HPackDecoder` level validation needs to know the order of the headers, we lose the order once we return Http2Headers and  `Http2HeadersValidation` needs to know more about the context: is this an initial header or trailing one? It didn't seem right to push that down to hpackdecoder.

Reply:
That's true regarding the order validation in hpack. Maybe we could shift request/response validation to where you're adding more validation logic, but probably not all of it and at a minimum this isn't obviously right.

Reply:
> If we use my strategy of having an observed variable for each critical header we should count no more than 4 as at 4 we'd hit an exceptional condition.

Thinking about it more, we don't have this guarantee since the headers order is not guaranteed in `Http2Headers`.
The implementation seems to keep the insertion order but this is not specified so potentially subject to change later.